{"cells": [{"cell_type": "code", "execution_count": null, "metadata": {}, "outputs": [], "source": ["#!/usr/bin/env python\n", "# coding: utf-8"]}, {"cell_type": "markdown", "metadata": {}, "source": ["In[1]:"]}, {"cell_type": "code", "execution_count": null, "metadata": {}, "outputs": [], "source": ["import os\n", "import sys\n", "from Bio import SeqIO\n", "import pandas as pd\n", "import seaborn as sns"]}, {"cell_type": "markdown", "metadata": {}, "source": ["inction to calculate and return gc content"]}, {"cell_type": "code", "execution_count": null, "metadata": {}, "outputs": [], "source": ["def getGC(seq):\n", "    return((sum([1.0 for nucl in seq if nucl in ['G', 'C']]) / len(seq)) * 100)"]}, {"cell_type": "markdown", "metadata": {}, "source": ["In[2]:"]}, {"cell_type": "markdown", "metadata": {}, "source": ["ead all the transcripts"]}, {"cell_type": "code", "execution_count": null, "metadata": {}, "outputs": [], "source": ["all_tx='../transcripts.fa'\n", "orphan_tx='../orphan_transcripts_final.fa'"]}, {"cell_type": "code", "execution_count": null, "metadata": {}, "outputs": [], "source": ["orphan_sequences = SeqIO.parse(open(orphan_tx),'fasta')\n", "orphan_ids=[]\n", "orphan_len=[]\n", "orphan_gc=[]\n", "for fasta in orphan_sequences:\n", "    name, sequence = fasta.id, str(fasta.seq)\n", "    orphan_ids.append(name)"]}, {"cell_type": "code", "execution_count": null, "metadata": {}, "outputs": [], "source": ["ids=[]\n", "lens=[]\n", "gc=[]\n", "types=[]\n", "all_sequences = SeqIO.parse(open(all_tx),'fasta')\n", "for fasta in all_sequences:\n", "    name, sequence = fasta.id, str(fasta.seq)\n", "    ids.append(name)\n", "    lens.append(len(sequence))\n", "    gc.append(getGC(sequence))\n", "    if name not in orphan_ids:\n", "        types.append('Non-orphan')\n", "    else:\n", "        types.append('Orphan')"]}, {"cell_type": "markdown", "metadata": {}, "source": ["In[3]:"]}, {"cell_type": "markdown", "metadata": {}, "source": ["et everything in pandas df"]}, {"cell_type": "code", "execution_count": null, "metadata": {}, "outputs": [], "source": ["nest=[ids,lens,gc,types]"]}, {"cell_type": "code", "execution_count": null, "metadata": {}, "outputs": [], "source": ["df=pd.DataFrame(nest, ['Name', 'len', 'gc','type']).T\n", "df[\"len\"]=pd.to_numeric(df[\"len\"])\n", "df[\"gc\"]=pd.to_numeric(df[\"gc\"])\n", "exon=pd.read_csv(\"exon_num.txt\",sep=\"\\t\")\n", "exon[\"number\"]=pd.to_numeric(exon[\"number\"])\n", "df=df.merge(exon,on=\"Name\",how=\"left\")"]}, {"cell_type": "markdown", "metadata": {}, "source": ["et all file names"]}, {"cell_type": "code", "execution_count": null, "metadata": {}, "outputs": [], "source": ["from os import listdir\n", "from os.path import isfile, join\n", "dirpath='../quant'\n", "quantfiles = [f for f in listdir(dirpath) if isfile(join(dirpath, f))]\n", "for f in quantfiles:\n", "    thisName=f.split(\"_\")[0]\n", "    fname=join(dirpath,f)\n", "    print(\"reading:\"+fname)\n", "    q1=pd.read_csv(fname,sep='\\t')\n", "    #print(q1)\n", "    thisQ=q1.loc[:,['Name','TPM']]\n", "    thisQ[\"TPM\"]=pd.to_numeric(thisQ[\"TPM\"])\n", "    thisQ.rename(columns = {'TPM':thisName+'_TPM'}, inplace = True) \n", "    #print(thisQ)\n", "    #join with original df\n", "    df=df.merge(thisQ,on='Name',how='left')"]}, {"cell_type": "markdown", "metadata": {}, "source": ["In[5]:"]}, {"cell_type": "markdown", "metadata": {}, "source": ["ompare length"]}, {"cell_type": "code", "execution_count": null, "metadata": {}, "outputs": [], "source": ["sns.set(font_scale=4)\n", "sns.set(style=\"whitegrid\")\n", "ax = sns.violinplot(x=\"type\", y=\"len\", data=df)\n", "ax.set_yscale(\"log\")\n", "ax.set_xlabel('Transcript age')\n", "ax.set_ylabel('Length')"]}, {"cell_type": "markdown", "metadata": {}, "source": ["In[6]:"]}, {"cell_type": "markdown", "metadata": {}, "source": ["ompare gc"]}, {"cell_type": "code", "execution_count": null, "metadata": {}, "outputs": [], "source": ["ax = sns.violinplot(x=\"type\", y=\"gc\", data=df)\n", "ax.set_xlabel('Transcript age');\n", "ax.set_ylabel('GC%');"]}, {"cell_type": "markdown", "metadata": {}, "source": ["In[7]:"]}, {"cell_type": "code", "execution_count": null, "metadata": {}, "outputs": [], "source": ["data=pd.melt(df.loc[:, 'type':'SRR999058_TPM' ],'type')\n", "import matplotlib.pylab as plt\n", "ax=sns.boxplot(x=\"variable\", y=\"value\",hue=\"type\", data=data)\n", "ax.set_yscale(\"log\")\n", "ax.set_xlabel('RNA-Seq Runs')\n", "ax.set_ylabel('Expression (TPM)')\n", "plt.xticks(rotation=45)"]}, {"cell_type": "markdown", "metadata": {}, "source": ["In[ ]:"]}], "metadata": {"kernelspec": {"display_name": "Python 3", "language": "python", "name": "python3"}, "language_info": {"codemirror_mode": {"name": "ipython", "version": 3}, "file_extension": ".py", "mimetype": "text/x-python", "name": "python", "nbconvert_exporter": "python", "pygments_lexer": "ipython3", "version": "3.6.4"}}, "nbformat": 4, "nbformat_minor": 2}(pyrpipe_human) [jingli@bigram2 downstream_analysis]
